packages/common/src/converter/data-converter.ts

```diff
@@ -182,8 +188,21 @@ export function mapToPayloadsSync<K extends string>(
   ) as Record<K, Payload>;
 }
 
-export const defaultDataConverter = new CompositeDataConverter(
-  new UndefinedPayloadConverter(),
-  new BinaryPayloadConverter(),
-  new JsonPayloadConverter()
-);
+export interface DefaultDataConverterOptions {
+  protobufClasses?: Record<string, unknown>;
```

Comment:
This is a bit unfortunate that the data converter doesn't accept the type/object to (de)serialize to at (de)serialize time. This is something other SDKs provide which means the implicit default data converter works with protos. Does this mean in order to use protos in TS, I have to manually configure a data converter?

One wonders if there is value in sort of proto converter proxy. For example, if the `encoding` is `json/protobuf` or `binary/protobuf`, can `fromPayload` return a `class ProtoMessage { #raw: Uint8Array; toProto<T>(obj: T); }` which then doesn't require such preconfiguration?

Reply:
> This is a bit unfortunate that the data converter doesn't accept the type/object to (de)serialize to at (de)serialize time

When encoding, the data converter gets the object, and we could just use that (something like `(arg.constructor as unknown as Encodable).encode(args).finish()`), but we need the classes and `metadata.messageType` in order to decode.

> Does this mean in order to use protos in TS, I have to manually configure a data converter?

Yes

> One wonders if there is value in sort of proto converter proxy. For example, if the encoding is json/protobuf or binary/protobuf, can fromPayload return a class ProtoMessage { #raw: Uint8Array; toProto<T>(obj: T); } which then doesn't require such reconfiguration?

The pro would be no config needed. The con is users would have to call `toProto` a lot. As a user, I'd personally prefer to do the config once than to write the `toProto` code. Current:

```ts
import type { ProtoActivityInput } from './protos'

export async function protoActivity(args: ProtoActivityInput): Promise<ProtoActivityResult> {
  return new ProtoActivityResult({ sentence: `${args.name} is ${args.age} years old.` });
}
```

With `toProto`:

```ts
import { ProtoActivityInput } from './protos'

export async function protoActivity(message: ProtoMessage): Promise<ProtoActivityResult> {
  const args = message.toProto<ProtoActivityInput>(ProtoActivityInput)
  return new ProtoActivityResult({ sentence: `${args.name} is ${args.age} years old.` });
}
```

It's only one line here, but it's a lot more code when you consider multiple arguments, everywhere arguments or return values are deserialized.

Reply:
I am not sure we should require other SDKs to support message type or users in this SDK to change the default converter because reification/reflection is not available. I wonder if instead you should have to be more explicit at runtime about the contract of the activity in cases where the types are ambiguous.

I would have to think a bit on how one can specify the proto classes for activity/workflow arguments at runtime leveraging the existing API without infecting payloads for everyone.

I agree that the `toProto` does look ugly that way. I was thinking about explicit data converter use, but yes for implicit conversion like for activity params, a better-defined contract is needed.

Having said this, this will continue to come up for other dynamically typed languages. We have to decide for dynamically typed languages whether the sender defines the proto type or the recipient.

Reply:
I am starting to formulate an idea...

What if _all_ TS protos had to be (google.protobuf.Any](https://developers.google.com/protocol-buffers/docs/proto3#any)? [Here](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto)'s the raw proto which is basically what you're doing here. They have [well defined JSON semantics](https://developers.google.com/protocol-buffers/docs/proto3#json) that makes them still look good when viewed (basically just `{ "@type": "type", ...restOfProto }`). They are also very cross-language compatible.

This can all be transparent and only affects cross-language compat. So basically the data converter converts every proto to an `Any`, and only accepts `Any` input. This would force other languages to either accept `Any` if executing something from TS, or send out `Any` if communicating with TS, which gets the built-in type name. Though that's still a bit gross compared to potentially changing all of them to use a new metadata header.

Well, now I can't find a global registry for `google.protobuf.Any` in protobufjs. Now I am starting to lean back to your original suggestion of putting `protoMessageType` as a metadata value in the other SDKs. Ignore my rant, heh (but I'll leave it here).

Reply:
> We went in this direction because I feel like it gives the best experience in JS / TS 

It is unfortunate that this experience though is going to affect other SDKs. It is unfortunate but I guess it's the best. I haven't spent too much time deriving an alternative.